src/overseer/store/datomic.clj

```diff
@@ -0,0 +1,128 @@
+(ns ^:no-doc overseer.store.datomic
+  "Internal core functions"
+  (:require [datomic.api :as d]
+            [overseer.core :as overseer]
+            [clojure.set :as set]
+            (loom graph)))
+
+(defn transitive-dependents [db job-id]
+  "Returns a set of job IDs that transitively depend upon given job ID
+   Basically recursive breadth-first graph traversal of the job graph."
+  (let [rules '[[(dependent ?j1 ?j0)
+                 [?j1 :job/dep ?j0]]
+                [(dependent ?j1 ?j0)
+                 [?j1 :job/dep ?j2]
+                 (dependent ?j2 ?j0)]]]
+    (set (d/q '[:find [?dep-jid ...]
+                :in $ % ?jid
+                :where [?j0 :job/id ?jid]
+                       [dependent ?j1 ?j0]
+                       [?j1 :job/id ?dep-jid]]
+              db
+              rules
+              job-id))))
+
+(defn jobs-ready'
+  "Find all job IDs that are ready to run, i.e. :unstarted
+  and not blocked (i.e. dependent on an unfinished job)"
+  [db]
+  (let [rules '[[(blocked ?j)
+                 [?j :job/dep ?dep]
+                 [?dep :job/status ?js]
+                 [(not= :finished ?js)]]]]
+    (->> (d/q '[:find [?jid ...]
+                :in $ %
+                :where [?j :job/status :unstarted]
+                       (not [blocked ?j])
+                       [?j :job/id ?jid]]
+              db
+              rules)
+         set)))
+
+(defn jobs-dead' [db thresh]
+  (d/q '[:find [(pull ?e [:*]) ...]
+         :in $ ?thresh
+         :where [?e :job/status :started]
+                [?e :job/id ?jid]
+                [?e :job/heartbeat ?h]
+                [(< ?h ?thresh)]]
+       db
+       thresh))
+
+(defn loom-graph->datomic-txn
+  "Given a Loom job graph, return a Datomic transaction representing it."
+  [graph]
+  (let [dep-edge
+        (fn [[job0 job1]]
+          [:db/add [:job/id (:job/id job0)] :job/dep [:job/id (:job/id job1)]]) ]
+    (concat
+      (map #(assoc % :db/id (d/tempid :db.part/user)) (loom.graph/nodes graph))
+      (map dep-edge (loom.graph/edges graph)))))
+
+(defn heartbeat []
+  (quot (System/currentTimeMillis) 1000))
```

Comment:
If stuff is in terms of UNIX timestamps, then this logicy stuff should be common and _outside_ the protocol implementation, so it can be shared _across_ implementations that all implement the protocol, not _inside_ the Datomic-specific implementation.